text/0080-datetime-extension.md

```diff
@@ -0,0 +1,367 @@
+Unix timestamps fail the readability criteria: an expression like `context.now < 1723000000` is difficult to read and understand without additional tooling to decode the timestamp.
+Unix timestamps are also indistinguishable from any other integral value providing no additional help in expressing the abstraction of time.
+Passing in pre-computed values in the context (e.g., a field like `context.isWorkingHours`) makes authorization logic difficult to audit because it moves this logic outside of Cedar policies and into the calling application.
+It is also difficult for the calling application to predict the necessary pre-computed values that a policy writer requires for their intended purpose. These may change over time, and may also differ depending on the principal, action, and/or resource.
+
+### Additional examples
+
+**Only allow user "alice" to view JPEG photos for one week after creation.**
+
+```cedar
+permit(
+  principal == User::"alice",
+  action == PhotoOp::"view",
+  resource is Photo
+) when {
+  resource.fileType == "JPEG" &&
+  (context.now.timestamp - resource.creationTime) <= duration(7, "days")
+};
+```
+
+**Allow access from a certain IP address only between 9am and 6pm UTC.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  context.workdayStart <= context.now.timestamp &&
+  context.now.timestamp < context.workdayEnd
+};
+```
+
+**Allow access from a certain IP address only between 9am and 6pm localized to their time.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"access",
+  resource
+) when {
+  context.srcIp.isInRange(ip("192.168.1.0/24")) &&
+  context.workdayStart <= context.now.timestamp &&
+  context.now.timestamp < context.workdayEnd
+};
+```
+
+Note that this is identical to the previous policy. The calling application must adjust `workdayStart` and `workdayEnd`.
+
+**Prevent employees from accessing work documents on the weekend.**
+
+```cedar
+forbid(
+  principal,
+  action == Action::"access",
+  resource is Document
+) when {
+  [1, 7].context.now.dayOfWeek
+};
+```
+
+**Permit access to a special opportunity for persons born on Leap Day.**
+
+```cedar
+permit(
+  principal,
+  action == Action::"redeem",
+  resource is Prize
+) when {
+  principal.birthDate.day == 29 &&
+  principal.birthDate.month == 2 &&
+  context.now.day == 29 && 
+  context.now.month == 2
+};
+```
+
+**Forbid access to EU resources after Brexit**
+
+```cedar
+forbid(
+  principal,
+  action,
+  resource
+) when {
+  context.now.timestamp > datetime("20200131T23:00:00Z") &&
+  context.location.countryOfOrigin == 'GB' &&
+  resource.owner == 'EU'
+}
+```
+
+## Detailed design
+
+This RFC proposes supporting two new extension types: `datetime`, which represents a particular instant of time, up to millisecond accuracy, and `duration` which represents a duration of time.
```

Comment:
Why millisecond accuracy? Why not more or less? I wonder about precedents, use-cases, etc. that justify this.

Reply:
https://github.com/cedar-policy/rfcs/pull/80#discussion_r1718463176

Reply:
Why not nanoseconds? That's what OPA/Rego supports. Also: What about other Date/Time libraries, e.g., available from Python.

Reply:
I'd be happy with nanoseconds if you can think of a use case for them.

Reply:
*I* don't have a use case for nanoseconds (and probably not millis, either, for that matter), but if you do please present it so we can discuss. 

Similarly, if someone has a use case for Coordinated Mars Time, instead of Unix Time (NASA... any interest in Cedar?), let's discuss. :)

Reply:
For milliseconds I can imagine something like "allow at most one access per 100ms" to prevent scraping. So `context.current_time > context.last_request_time + duration(100, "ms")`. I can't think of a use for nanoseconds.

Reply:
Update: we've added a section on operator accuracy to the "unresolved questions". Please move conversation to a comment on that section!